File: problem69/Problem69.java
import java.util.*;
/**
 * Problem 69
 */

public class Problem69 {
	public static void main(String[] arg) {
		int n = (int)Math.sqrt(1000000);
		boolean[] prime = IntegerArithmetic.sieveEratosthenes(n);
		int nbPrimes = 0;
		for (int i = 0; i < prime.length; i++) {
			if (prime[i]) nbPrimes++;
		}
		int[] primes = new int[nbPrimes];
		int j = 0;
		for (int i = 0; i < prime.length; i++) {
			if (prime[i]) {
				primes[j] = i;
				j++;
			}
		}
		float maxRatio = 0;
		int index = 0;
		for (int i = 2; i <= 1000000; i++) {
			float ratio = (float)i / (float)phi(i, primes);
			if (ratio > maxRatio) {
				maxRatio = ratio;
				index = i;
			} 
		}
		System.out.println("The solution is " + index);
	}

	/**
	* Returns phi n
	*/
	public static int phi(int n, int[] primes) {
		ArrayList<ArrayList<Integer>> primeDec = IntegerArithmetic.primeDecomp(n, primes);
		int result = 1;
		for (int i = 0; i < primeDec.size(); i++) {
			result *= (primeDec.get(i).get(0) - 1) * (int)Math.pow(primeDec.get(i).get(0), primeDec.get(i).get(1) - 1);
		}
		return result;
	}
}

File: problem69/problem69.js
/**
 * Problem 69
 */

var IntegerArithmetic = require('../libraries/IntegerArithmetic.js');
var n = Math.floor(Math.sqrt(1000000));
var prime = IntegerArithmetic.sieveEratosthenes(n);
var nbPrimes = 0;
for (var i = 0; i < prime.length; i++) {
	if (prime[i]) nbPrimes++;
}
var primes = [];
var j = 0;
for (var i = 0; i < prime.length; i++) {
	if (prime[i]) primes.push(i);
}
var maxRatio = 0;
var index = 0;
for (var i = 2; i <= 1000000; i++) {
	var ratio = i / phi(i, primes);
	if (ratio > maxRatio) {
		maxRatio = ratio;
		index = i;
	} 
}
console.log ('The solution is ' + index);

/**
* Returns phi n
* @param integer n
* @param array of integers primes
* @return integer
*/
function phi(n, primes) {
	var primeDec = IntegerArithmetic.primeDecomp(n, primes);
	var result = 1;
	for (var i = 0; i < primeDec.length; i++) {
		result *= (primeDec[i][0] - 1) * Math.floor(Math.pow(primeDec[i][0], primeDec[i][1] - 1));
	}
	return result;
}